pyclashbot/interface/ui.py

```diff
@@ -0,0 +1,799 @@
+        for key, var in self.gp_vars.items():
+            widget = self._config_widgets.get(key)
+            if not widget:
+                continue
+            values = [str(option) for option in widget.cget("values")]
+            if var.get() not in values and values:
+                var.set(values[0])
+
+    def _apply_theme(self, theme_name: str, skip_variable_update: bool = False) -> None:
+        available = tuple(self._style.theme_names())
+        selected = theme_name if theme_name in available else self.DEFAULT_THEME
+        if selected not in available and available:
+            selected = available[0]
+        if not skip_variable_update or self.theme_var.get() != selected:
+            self._suspend_traces += 1
+            try:
+                self.theme_var.set(selected)
+            finally:
+                self._suspend_traces -= 1
+        self._style.theme_use(selected)
+        self._refresh_theme_colours()
+
+    def _label_foreground(self) -> str:
+        colour = self._style.lookup("TLabel", "foreground")
+        return colour or "#202020"
+
+    def _refresh_theme_colours(self) -> None:
+        foreground = self._label_foreground()
+        for label in self._theme_labels:
+            try:
+                label.configure(foreground=foreground)
+            except tk.TclError:
+                continue
+        gauge_fg = getattr(self._style.colors, "success", "#2ecc71") if hasattr(self._style, "colors") else "#2ecc71"
+        gauge_bg = getattr(self._style.colors, "danger", "#e74c3c") if hasattr(self._style, "colors") else "#e74c3c"
+        self.win_gauge.set_colours(gauge_fg, gauge_bg, foreground)
+
+    def _on_theme_change(self, _event: object) -> None:
+        self._apply_theme(self.theme_var.get(), skip_variable_update=True)
+        self._notify_config_change()
+
+    def _on_emulator_changed(self) -> None:
+        self._show_emulator_settings()
+        self._notify_config_change()
+
+    def _show_emulator_settings(self) -> None:
+        index_map = {"Google Play": 0, "MEmu": 1, "BlueStacks 5": 2}
+        self.sub_notebook.select(index_map.get(self.emulator_var.get(), 1))
+
+    def _hide_action_button(self) -> None:
+        self.action_btn.grid_remove()
+        self.stop_btn.grid()
+
+    def _on_action_pressed(self) -> None:
+        if self._action_callback:
+            self._action_callback()
+        self._hide_action_button()
+
+    def _on_open_recordings_clicked(self) -> None:
+        if self._open_recordings_callback:
+            self._open_recordings_callback()
+
+    def _on_open_logs_clicked(self) -> None:
+        if self._open_logs_callback:
+            self._open_logs_callback()
+
+    @staticmethod
+    def _safe_int(value: object, fallback: int = 0) -> int:
+        try:
+            return int(value)
+        except (TypeError, ValueError):
+            return fallback
+
+    @staticmethod
+    def _extract_winrate(raw: object, wins: int, losses: int) -> float:
```

Comment:
this is confusing...
function takes in an object which can be a string in some cases, which it then parses into a float
OR
it uses the wins and losses passed to calculate a win rate.

this seems like we need two functions for two different cases, can we refactor to have single responsibility? 

im also not a huge fan of the `object` type, if we know what it should be, then type it.